### PR TITLE
Add sentence_transformer example with xnnpack and cpu backend

### PR DESCRIPTION
**Pull request by seyeong-han**
# Sentence Transformers for ExecuTorch

## Summary

This PR adds comprehensive support for exporting and running **sentence-transformer models** with ExecuTorch, enabling high-performance semantic embeddings on edge devices.

## Key Features

- ✅ **76x performance improvement** with XNNPack backend (15.9ms vs 1,212ms on M3 Pro)
- ✅ **Perfect accuracy** - Cosine similarity = 1.0 with original HuggingFace models
- ✅ **Generic framework** - Works with any encoder-based sentence-transformer model
- ✅ **Production-ready tooling** - Export, validation, benchmarking, and C++ deployment
- ✅ **Complete documentation** - README, usage examples, and performance comparisons

## Files Added

```
examples/models/sentence_transformer/
├── model.py                              # Model wrapper with mean pooling
├── export_sentence_transformer.py        # Export script (XNNPack/CPU backends)
├── compare_embeddings.py                 # Embedding validation tool
├── benchmark_backends.py                 # Performance benchmarking (XNNPack vs CPU)
├── create_input_bins.py                  # Binary input generation for C++ runtime
├── install_requirements.sh               # Dependency installation
└── README.md                             # Comprehensive documentation

.ci/scripts/
└── test_sentence_transformer.sh          # CI validation tests

Root level:
├── CMakePresets.json                     # Added executor-runner-xnnpack preset
└── Makefile                              # Added executor-runner-xnnpack target
```

## Tested Models

All models verified with **perfect embedding quality** (cosine similarity = 1.0):

| Model                 | Params | Embedding Dim | Export | Inference | Quality            |
| --------------------- | ------ | ------------- | ------ | --------- | ------------------ |
| **all-MiniLM-L6-v2**  | 22.7M  | 384           | ✅     | ✅        | Perfect (1.000000) |
| **all-MiniLM-L12-v2** | 33.4M  | 384           | ✅     | ✅        | Perfect (1.000000) |
| **all-mpnet-base-v2** | 109M   | 768           | ✅     | ✅        | Perfect (1.000000) |

## Performance Results

Benchmarked on **Apple M3 Pro** with all-MiniLM-L6-v2:

| Backend                  | Average Latency | Throughput       | Speedup         |
| ------------------------ | --------------- | ---------------- | --------------- |
| **ExecuTorch XNNPack**   | **15.9 ms**     | **~63 sent/sec** | **76x**       |
| ExecuTorch CPU           | 1,212 ms        | ~0.83 sent/sec   | 1.0x (baseline) |
| HuggingFace Transformers | 9.5 ms          | ~110 sent/sec    | -               |

**Key Insights:**

- XNNPack provides **76x speedup** over CPU backend for edge deployment
- HuggingFace is faster on desktop due to PyTorch optimizations, but ExecuTorch enables mobile/embedded deployment
- Perfect embedding accuracy maintained across all backends

### Export Model

```bash
cd examples/models/sentence_transformer

# Export with XNNPack (recommended)
python export_sentence_transformer.py \
    --model sentence-transformers/all-MiniLM-L6-v2 \
    --backend xnnpack \
    --output-dir ./xnnpack_model
```

### Validate Embeddings

```bash
# Verify exported model matches original
python compare_embeddings.py \
    --model-path ./xnnpack_model/model.pte \
    --model-name sentence-transformers/all-MiniLM-L6-v2 \
    --sentences "Test sentence for validation"
```

### Benchmark Performance

```bash
# Compare XNNPack vs CPU performance
python benchmark_backends.py --iterations 100
```

### C++ Runtime Deployment

```bash
# Build executor_runner with XNNPack
make executor-runner-xnnpack

# Create binary inputs
python create_input_bins.py --text "Your text here"

# Run inference
./cmake-out/executor_runner \
    --model_path=./xnnpack_model/model.pte \
    --inputs=input_ids.bin,attention_mask.bin
```

## Use Cases Enabled

- **Mobile semantic search** - Documents, contacts, emails
- **On-device recommendations** - Content similarity without cloud
- **Privacy-preserving matching** - All processing on-device
- **Edge AI applications** - IoT, embedded systems, offline scenarios
- **RAG on mobile** - Retrieval-augmented generation for edge devices

### Model Architecture

```
Input Text
    ↓
Tokenization → [input_ids, attention_mask]
    ↓
Transformer Encoder → [batch, seq_len, hidden_size]
    ↓
Mean Pooling (attention-mask weighted)
    ↓
Sentence Embedding → [batch, embedding_dim]
```

### Backend Support

- **XNNPack** (Recommended) - Optimized for ARM/x86 CPU inference
- **CPU** (Baseline) - Portable reference implementation

### Validation Methodology

1. Export model to ExecuTorch format
2. Run inference on same inputs with both original and exported models
3. Compare embeddings using:
   - Cosine similarity (target: 1.0)
   - L2 distance (target: ~0)
   - Mean/Max absolute error

All tested models achieve cosine similarity = 1.0 (numerically identical at ~1e-6 precision).

## CI Coverage

The CI script (`test_sentence_transformer.sh`) validates:

1. ✅ XNNPack export functionality
2. ✅ CPU export functionality
3. ✅ XNNPack embedding accuracy
4. ✅ CPU embedding accuracy
5. ✅ Benchmark tool execution
6. ✅ Binary input generation for C++ runtime

## Documentation

Comprehensive README includes:

- Model architecture explanation
- Installation instructions
- Export/validation/benchmarking workflows
- Python and C++ usage examples
- Performance comparison tables
- Troubleshooting guide
- Production deployment guidelines

## Testing

```bash
# Run CI tests
./.ci/scripts/test_sentence_transformer.sh

# Manual validation
cd examples/models/sentence_transformer
python export_sentence_transformer.py --backend xnnpack
python compare_embeddings.py --model-path sentence_transformer_export/model.pte
```

## Future Work

- [ ] Dynamic batching support
- [ ] INT8/INT4 quantization for smaller models
- [ ] Metal backend for Apple Silicon optimization
- [ ] CoreML backend for iOS integration
- [ ] Custom C++ runner example for mobile apps
- [ ] Additional model architectures (e.g., cross-encoders)

### CODE REVIEW

**Comment by seyeong-han** [comment]
@lucylq , @jackzhxng 

Is it okay to implement a Python runner for the initial code, or is it necessary to develop the C++ runner?

**Comment by seyeong-han** [comment]
> @lucylq , @jackzhxng
> 
> Is it okay to implement a Python runner for the initial code, or is it necessary to develop the C++ runner?

@lucylq @jackzhxng 

I've added the `executor-runner-xnnpack` runner for sentence-transformers models!
